Firestore/core/src/firebase/firestore/local/query_data.cc

```diff
@@ -78,25 +81,36 @@ QueryData::QueryData(Query query,
                 sequence_number,
                 purpose,
                 SnapshotVersion::None(),
+                SnapshotVersion::None(),
                 ByteString()) {
 }
 
 QueryData QueryData::Invalid() {
   return QueryData(Query(), /*target_id=*/-1, /*sequence_number=*/-1,
                    QueryPurpose::Listen,
+                   SnapshotVersion(SnapshotVersion::None()),
                    SnapshotVersion(SnapshotVersion::None()), {});
 }
 
 QueryData QueryData::WithSequenceNumber(
     ListenSequenceNumber sequence_number) const {
   return QueryData(Query(query_), target_id_, sequence_number, purpose_,
-                   snapshot_version_, resume_token_);
+                   snapshot_version_, last_limbo_free_snapshot_version_,
+                   resume_token_);
 }
 
 QueryData QueryData::WithResumeToken(ByteString resume_token,
                                      SnapshotVersion snapshot_version) const {
   return QueryData(Query(query_), target_id_, sequence_number_, purpose_,
-                   std::move(snapshot_version), std::move(resume_token));
+                   std::move(snapshot_version),
+                   last_limbo_free_snapshot_version_, std::move(resume_token));
+}
+
+QueryData QueryData::WithLastLimboFreeSnapshotVersion(
```

Comment:
Port of https://github.com/firebase/firebase-android-sdk/blob/2827026bafd7869f35c6c91cd4a2e9949c1fba7f/firebase-firestore/src/main/java/com/google/firebase/firestore/local/QueryData.java#L103